vhdbuilder/prefetch/pkg/container/container.go

```diff
@@ -0,0 +1,68 @@
+package container
+
+import (
+	"bytes"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"os"
+	"strings"
+)
+
+var (
+	//go:embed templates/prefetch.sh.gtpl
+	prefetchScript         string
+	prefetchScriptTemplate = template.Must(template.New("containerimageprefetch").Parse(prefetchScript))
+)
+
+// ParseComponents parses the named component list JSON and returns its content as a ComponentList.
+func ParseComponents(name string) (*ComponentList, error) {
+	bytes, err := os.ReadFile(name)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read component list %s: %w", name, err)
+	}
+
+	var list ComponentList
+	if err = json.Unmarshal(bytes, &list); err != nil {
+		return nil, fmt.Errorf("unable to unnmarshal component list content: %w", err)
+	}
+
+	if len(list.Images) < 1 {
+		return nil, fmt.Errorf("parsed list of container images from %s is empty", name)
+	}
+
+	return &list, nil
+}
+
+// Generate generates and saves the container image prefetch script to disk based on the specified component list and destination path.
+func Generate(components *ComponentList, dest string) error {
+	var args TemplateArgs
+	for _, image := range components.Images {
+		if len(image.PrefetchOptimizations) > 0 {
+			for _, optimization := range image.PrefetchOptimizations {
+				args.Images = append(args.Images, TemplateImage{
+					FullyQualifiedTag: fmt.Sprintf("%s%s", strings.TrimSuffix(image.DownloadURL, "*"), optimization.Tag),
```

Comment:
Does anything validate that `DownloadURL` always ends with "*"? (guessing the answer is "yes" since everything in components.json seem to have it)

Reply:
I actually originally thought that the cue schema did something along these lines, but this seems to not be the case - added an appropriate failure mode

technically if it wasn't formatted in this way, the container image pull steps within install-dependencies.sh would most likely fail: https://github.com/Azure/AgentBaker/blob/cameissner/cni-prefetch/vhdbuilder/packer/install-dependencies.sh#L237-L267